### ISSUE TITLE
Broadcasted bias with DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK

### ISSUE THREAD

**Issue by atafra** [ROCmSoftwarePlatform/composable_kernel#601]
Hi,

I need a 2D convolution + bias + ReLU kernel on Radeon GPUs (gfx1030). It seems to me that DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK is the right kernel for this. I have a 1D bias vector so this needs to be broadcasted to the output shape. The problem is that I can't get this to work with this kernel.

I initially tried to use GNHWK layout for the bias (D) and set the strides to broadcast it (0, 0, 1, 0, 0). But this results in invalid memory accesses. From the code it seems that GNHWK is a packed-only layout so the strides are ignored, and this explains the crash.

I then tried the strided G_NHW_K layout but that resulted in a compile error. Based on one of the included client examples (44_conv2d_fwd_quantization), it looks like the right layout for such bias vectors should be G_K or perhaps GK. But neither of these work, I get the same compile error. The example uses a different kernel though (DeviceGroupedConvFwdMultipleD_Xdl_CShuffle).

So I'm stuck at this point. Does the DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK kernel support 1D or strided layouts at all? If it doesn't, is there any alternative which supports gfx1030? Thanks!

Regards,
Attila

**Comment by junliume** [comment]
Ping @zjing14

**Comment by zjing14** [comment]
If I understand your question correctly. for GNHWK layout, the bias tensor should be length{1, 1, 1, 1, K} and stride{0, 0, 0, 0, 1}.

**Comment by atafra** [comment]
My main question is that should this particular kernel work with 1D bias in some way at all? Was this ever tested? I couldn't find any case in the entire repository where this kernel is used with 1D bias. Actually, in the `library` this kernel is not instantiated with any bias at all: https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/e6cda9f8ff8baa58d61905239a70f93db1933eb5/library/src/tensor_operation_instance/gpu/grouped_conv2d_fwd/device_grouped_conv2d_fwd_dl_gnhwc_gkyxc_gnhwk_f16_instance.cpp#L51

My bias tensor is packed as a 1D tensor but as a GNHWK tensor it cannot be considered packed because it's a 1D tensor broadcasted to a 4D one using custom strides.

What's the difference between the GNHWK and G_NHW_K layouts? My understanding is that GNHWK is a packed layout, and it seems to me that the strides are simply ignored for this layout, so it doesn't matter what strides do I set for the bias. The strides are seemingly used only for G_NHW_K layout.

**Comment by atafra** [comment]
@zjing14 could you please briefly comment on this? Due to the lack of documentation it's very hard to figure this out just by looking at the code.

**Comment by junliume** [comment]
Also @asroy

**Comment by zjing14** [comment]
@atafra Yes, you are right. GNHWK is packed, G_NHW_K means strided. (https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/device/tensor_layout.hpp)

No matter if the bias tensor is packed/unpacked, then the stride should be (0, 0, x, 0, 0). The memory fault may be because the convolution case you specify is not supported. Could you post your convolution case here?

**Comment by atafra** [comment]
I managed to solve the problem but it was quite hard to figure out. First, the convolution case is definitely supported. Also, as I said before, the strides are completely ignored for the GNHWC / GNHWK layout, which can be seen here: https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp#L814

So the crash happened because the strides were ignored and thus out-of-bounds accesses were performed by the kernel. I was trying to use this particular tensor layout because the example code and the library instantiations for this kernel are using it as well.

But looking at this code, I found out that for the NHWGC / NHWGK layout the strides are indeed **not** ignored. So using this layout with a broadcasted bias works as expected. I couldn't find any mention of this behavior in the documentation, and it's quite confusing because both GNHWC and NHWGC layouts are claimed to be packed but one of them supports custom strides too, while the other does not, without throwing any errors. It would be great to mention this in the documentation and also add some runtime checks or at least asserts for the strides because silently ignoring them is very error-prone.
